src/Client/Connection.cpp

```diff
@@ -127,7 +127,27 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
 
             try
             {
-                socket->connect(*it, connection_timeout);
+                if (async_callback)
+                {
+                    socket->connectNB(*it);
+                    while (!socket->poll(0, Poco::Net::Socket::SELECT_READ | Poco::Net::Socket::SELECT_WRITE | Poco::Net::Socket::SELECT_ERROR))
```

Comment:
@Avogar correct me if I'm wrong, but AFAIU it will return `fd` if it is not ready (poll returns false) directly to the PipelineExecutor, and it will be monitored with epoll, and this is the idea, since otherwise it will wait here for the `timeout`, while it can do something else.

Reply:
> maybe here it also makes sense to pass timeout?

Why? We need to return from poll immidiately (so timeout is 0). Connection timeout will be checked outside in epoll

> @Avogar correct me if I'm wrong, but AFAIU it will return fd if it is not ready (poll returns false) directly to the PipelineExecutor, and it will be monitored with epoll, and this is the idea, since otherwise it will wait here for the timeout, while it can do something else.

Yes. If poll returns false we call async_callback and go back to PipelineExecutor (maybe first to epoll from HedgedConnectionsFactory and then this epoll goes to PipelineExecutor if hedged requests are used).